jlm/llvm/opt/cne.cpp

```diff
@@ -288,44 +288,43 @@ mark(rvsdg::Region *, cnectx &);
 static void
 mark_gamma(const rvsdg::StructuralNode * node, cnectx & ctx)
 {
-  JLM_ASSERT(rvsdg::is<rvsdg::GammaOperation>(node->GetOperation()));
+  auto & gamma = dynamic_cast<const rvsdg::GammaNode &>(*node);
 
   /* mark entry variables */
-  for (size_t i1 = 1; i1 < node->ninputs(); i1++)
+  for (size_t i1 = 1; i1 < gamma.ninputs(); i1++)
   {
-    for (size_t i2 = i1 + 1; i2 < node->ninputs(); i2++)
-      mark_arguments(node->input(i1), node->input(i2), ctx);
+    for (size_t i2 = i1 + 1; i2 < gamma.ninputs(); i2++)
+      mark_arguments(gamma.input(i1), gamma.input(i2), ctx);
   }
 
-  for (size_t n = 0; n < node->nsubregions(); n++)
-    mark(node->subregion(n), ctx);
+  for (size_t n = 0; n < gamma.nsubregions(); n++)
+    mark(gamma.subregion(n), ctx);
 
   /* mark exit variables */
-  for (size_t o1 = 0; o1 < node->noutputs(); o1++)
+  for (size_t o1 = 0; o1 < gamma.noutputs(); o1++)
   {
-    for (size_t o2 = o1 + 1; o2 < node->noutputs(); o2++)
+    for (size_t o2 = o1 + 1; o2 < gamma.noutputs(); o2++)
     {
-      if (congruent(node->output(o1), node->output(o2), ctx))
-        ctx.mark(node->output(o1), node->output(o2));
+      if (congruent(gamma.output(o1), gamma.output(o2), ctx))
+        ctx.mark(gamma.output(o1), gamma.output(o2));
     }
   }
 }
 
 static void
 mark_theta(const rvsdg::StructuralNode * node, cnectx & ctx)
 {
-  JLM_ASSERT(is<rvsdg::ThetaOperation>(node));
-  auto theta = static_cast<const rvsdg::ThetaNode *>(node);
+  auto & theta = dynamic_cast<const rvsdg::ThetaNode &>(*node);
 
   /* mark loop variables */
-  for (size_t i1 = 0; i1 < theta->ninputs(); i1++)
+  for (size_t i1 = 0; i1 < theta.ninputs(); i1++)
   {
-    for (size_t i2 = i1 + 1; i2 < theta->ninputs(); i2++)
+    for (size_t i2 = i1 + 1; i2 < theta.ninputs(); i2++)
     {
-      auto input1 = theta->input(i1);
-      auto input2 = theta->input(i2);
-      auto loopvar1 = theta->MapInputLoopVar(*input1);
-      auto loopvar2 = theta->MapInputLoopVar(*input2);
+      auto input1 = theta.input(i1);
+      auto input2 = theta.input(i2);
+      auto loopvar1 = theta.MapInputLoopVar(*input1);
+      auto loopvar2 = theta.MapInputLoopVar(*input2);
```

Comment:
Why are those changes here? They have nothing to do with the phi node.

Reply:
reverted all